exposure-notification/src/main/java/fi/thl/covid19/exposurenotification/FederationGatewayRestClientConfiguration.java

```diff
@@ -0,0 +1,101 @@
+package fi.thl.covid19.exposurenotification;
+
+import fi.thl.covid19.exposurenotification.error.RestTemplateErrorHandler;
+import org.apache.http.conn.ssl.SSLConnectionSocketFactory;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.ssl.PrivateKeyStrategy;
+import org.apache.http.ssl.SSLContextBuilder;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.web.client.RestTemplateBuilder;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
+import org.springframework.util.ResourceUtils;
+import org.springframework.web.client.RestTemplate;
+
+import javax.net.ssl.SSLContext;
+import java.io.File;
+import java.io.FileInputStream;
+import java.security.KeyStore;
+import java.time.Duration;
+
+@Configuration
+public class FederationGatewayRestClientConfiguration {
+
+    private final Duration connectTimeout = Duration.ofSeconds(60);
+    private final Duration readTimeout = Duration.ofSeconds(60);
+    private final Duration socketTimeout = Duration.ofSeconds(60);
+
+    private final String trustStorePath;
+    private final char[] trustStorePassword;
+    private final String keyStorePath;
+    private final char[] keyStorePassword;
+    private final String keyStoreKeyAlias;
+
+    public FederationGatewayRestClientConfiguration(
+            @Value("${covid19.federation-gateway.trust-store.path}") String trustStorePath,
+            @Value("${covid19.federation-gateway.trust-store.password}") String trustStorePassword,
+            @Value("${covid19.federation-gateway.client-key-store.path}") String keyStorePath,
+            @Value("${covid19.federation-gateway.client-key-store.password}") String keyStorePassword,
+            @Value("${covid19.federation-gateway.client-key-store.alias}") String keyStoreKeyAlias
+    ) {
+        this.trustStorePath = trustStorePath;
+        this.trustStorePassword = trustStorePassword.toCharArray();
+        this.keyStorePath = keyStorePath;
+        this.keyStorePassword = keyStorePassword.toCharArray();
+        this.keyStoreKeyAlias = keyStoreKeyAlias;
+    }
+
+    @Bean("federationGatewayRestTemplate")
+    public RestTemplate customRestTemplate(RestTemplateBuilder builder) {
+        return builder
+                .setConnectTimeout(connectTimeout)
+                .setReadTimeout(readTimeout)
+                .setConnectTimeout(socketTimeout)
+                .requestFactory(this::requestFactory)
+                .errorHandler(new RestTemplateErrorHandler())
+                .build();
+    }
+
+    private HttpComponentsClientHttpRequestFactory requestFactory() {
+        try {
+            SSLContext context = (!trustStorePath.isBlank() &&
```

Comment:
Yep, good point. I'll check this later.

Reply:
Actually, I checked this briefly and I don't quite get what you mean by making whole class optional. There is 4 values and yes, those can be bundled into configuration properties, but don't get how this solves anything. You could move the logic inside properties yes, but not so sure if this is really much better. Maybe I just don't get what the actual point in here is.

Reply:
Anyway, I moved these to properties into property class and moved validation logic also inside that class. It's at least a little bit cleaner now.

Reply:
The point was that since you're using it as all-or-nothing, you can bundle the properties up and validate that each field inside exists and is good - the whole thing must be missing if anything is. Then in here, you only need to check whether you have a config or don't - it can never be invalid/partial so you don't need to check each field for length/blank (mid-states are in.

The Optional part is just styling on top of that. You can take the trustStorePath as an argument for createSslContext() rather than using it from the field (more explicit) and then this becomes a question of:
`context = config != null ? createSslContext(config) : SSLContextBuilder.create().build()` 
vs 
`config.map(this::createSslContext).orElseGet(this::createDefaultSslContext)`

Reply:
Are you sure that config can be missing. I think Spring Boot application won't start if configuration is not initialized properly. Anyway, I moved initialization state in the own method inside configuration properties.

Reply:
Not 100% as I haven't done it on a configuration bean, but I think so. There's a variety of ways to make a beans optional (conditionalonproperty for example), and I don't see why they wouldn't work on configuration beans as well. You can certainly take a dependency-bean in as optional.
That aside, your solution looks good too.

Reply:
According to docs, conditional configuration beans are a thing:
https://docs.spring.io/spring-boot/docs/1.2.1.RELEASE/reference/html/boot-features-developing-auto-configuration.html

Reply:
yep, only problem might be if conditional annotations can be used in ConfigurationProperties. I tried this briefly, with following annotations in property class:

`@ConditionalOnExpression("${covid19.federation-gateway.rest-client.trust-store.path} > ''")
@ConditionalOnProperty("${foo.bar} = true")`

Separately, not the same time. I expect to get null pointer exception with both, but everything runs smoothly, which indicates that conditional have no effect on property class. Anyway, this needs more investigation, maybe later .. :)

Reply:
Anyway, even if this approach don't work directly in properties, this strategy could be used. However, result might not be as simple as one might think on first look.